### PR TITLE
adds support for passing custom headers along side urls

### PR DESCRIPTION
**Pull request by mogorman**
When I first tried to use the library with the mbta api all the examples i saw involved passing in the api-key as a header and not as just another arg in the url string. So I added support in the library for custom headers. I then saw you can pass the key in the url and it wasn't needed for my project. regardless I thought it was a nice feature to add to your library.

### CODE REVIEW

File: lib/server_sent_event_stage.ex [hunk 1]
```diff
@@ -22,23 +22,25 @@ defmodule ServerSentEventStage do
   """
   def start_link(args) do
     url = Keyword.fetch!(args, :url)
+    headers = Keyword.get(args, :headers, [])
     opts = Keyword.take(args, ~w(debug name timeout spawn_opt)a)
-    GenStage.start_link(__MODULE__, url, opts)
+    GenStage.start_link(__MODULE__, %{url: url, headers: headers}, opts)
```
Comment (paulswartz) [inline lib/server_sent_event_stage.ex]: We're `fetch!`ing the URL more to ensure that it's there. It'll be simpler to pass the `args` directly into `init`:
```elixir
_url = Keyword.fetch!(args, :url)
opts = ...
GenStage.start_link(__MODULE__, args, opts)
```

File: lib/server_sent_event_stage.ex [hunk 2]
```diff
@@ -22,23 +22,25 @@ defmodule ServerSentEventStage do
...
   end
 
   # Server functions
-  defstruct [:url, buffer: "", state: :not_connected]
+  defstruct [:url, :headers, buffer: "", state: :not_connected]
 
   @doc false
-  def init(url) do
-    state = %__MODULE__{url: url}
+  def init(%{url: url, headers: headers}) do
```
Comment (paulswartz) [inline lib/server_sent_event_stage.ex]: Then, with `args` as the argument here:
```
def init(args) do
  state = %__MODULE__{
    url: Keyword.fetch!(args, :url),
    headers: Keyword.get(args, :headers, [])
  }
  {:producer, state}
end
```

File: lib/server_sent_event_stage.ex [hunk 3]
```diff
@@ -22,23 +22,25 @@ defmodule ServerSentEventStage do
...
+    state = %__MODULE__{url: url, headers: headers}
     {:producer, state}
   end
 
   @doc false
   def handle_info(:connect, state) do
     url = compute_url(state)
-    :ok = connect_to_url(url)
+    %{url: _url, headers: headers} = state
+    :ok = connect_to_url(url, headers)
```
Comment (paulswartz) [inline lib/server_sent_event_stage.ex]: You don't need to match out the headers here:
```elixir
url = compute_url(state)
:ok = connect_to_url(url, state.headers)
```

File: lib/server_sent_event_stage.ex [hunk 4]
```diff
@@ -97,7 +99,8 @@ defmodule ServerSentEventStage do
   end
 
   def handle_info(%HTTPoison.AsyncEnd{}, %{state: {:redirect, new_url}} = state) do
-    :ok = connect_to_url(new_url)
+    %{url: _url, headers: headers} = state
+    :ok = connect_to_url(new_url, headers)
```
Comment (paulswartz) [inline lib/server_sent_event_stage.ex]: Same here: you don't need to match out the headers.

**Review by paulswartz — COMMENTED** [review]
Thanks for starting this!

Could you also add a test that shows the headers being passed to the server? You should be able to add a `test` to the "bypass" section in server_sent_event_test.exs.

**Comment by mogorman** [comment]
No problem. Added positive and negative unit test, and made changes you described.

**Review by paulswartz — APPROVED** [review]
Thanks again! 🍰

**Pull request merged**